### PR TITLE
[WIP, don't merge] Option refurbish

### PR DESCRIPTION
**Pull request by glennsl**
Names I'm unsure about:
* `flatMap` vs `andThen`
* `and_` vs `then_`
* `flatten` vs `join`
* `zip` vs `both`
* `make` vs `return` (it's really just an alias for `Some _` and `make` seems too general for monads like `result` with multiple type vars)
* `get` as an alias for `getOr` without argument label
* `maybe` as alias for `mapOr` without argument label
* `getOrRaise` vs `getExn` or similar
* `okOr` vs `toResult`
* `reduce` vs `fold`. Some libraries seem to have both, where `fold` does not take an initial value but instead raises an exception on `None`. So reduce seems slightly more correct, and is of course much more familiar to js devs.

### CODE REVIEW

**Comment by glennsl** [comment]
I'm still unsure about the naming convention used, there are no tests yet, and the docs are pretty half-baked. But feedback on what's there is appreciated nonetheless :)

File: package.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
   "description": "This is a bs port of the ocaml-containers project",
   "main": "index.js",
   "scripts": {
+    "start": "bsb -w",
```
Comment (dorafmon) [inline package.json]: What's the convention here, `npm start` or `npm watch?`

Comment (glennsl) [inline package.json]: `start` is an actual `npm` command while watch is not, you have to do `npm run watch`, so I prefer `start`

**Comment by glennsl** [comment]
@bordoley Isn't that just `map2`?

**Comment by bordoley** [comment]
Ah i didn't realized you had both.

**Comment by glennsl** [comment]
@dorafmon I've now added tests, so this should be good to merge if you approve. I'm still unsure about the naming though.

I've also tweaked the test setup to be a bit more convenient to use.

File: .screenrc [hunk 2]
```diff
@@ -0,0 +1,5 @@
+sessionname "bs-containers test console"
```
Comment (dorafmon) [inline .screenrc]: I think we should ignore this?

File: test/__tests__/bopt_test.ml [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    deepEqual t (getLazy (fun () -> "b") None) "b");
+
+  test "forEach - Some _" (fun t ->
+    plan t 1;
+    forEach (fun x -> deepEqual t x "a") (Some "a"));
+  test "forEach - None" (fun t ->
+    forEach (fun _ -> fail t) None);
+
+  test "map - Some _" (fun t ->
+    deepEqual t (map (( *) 3) (Some 7)) (Some 21));
```
Comment (dorafmon) [inline test/__tests__/bopt_test.ml]: Also we are getting this warning, do you know why? `Warning 2: this is not the end of a comment.`

Comment (glennsl) [inline test/__tests__/bopt_test.ml]: OCaml includes a syntactic linter that is incapable of distinguishing good code from bad code, in this case end comment vs multiplication operator. Fixed in the following commit.

Comment (dorafmon) [inline test/__tests__/bopt_test.ml]: Thanks! Should we remove the .screenrc as well, I think that is your local setting?

File: .screenrc (hunk 2 above)
Comment (glennsl) [inline .screenrc]: The whole file? That's intentional. It's the configuration file for GNU screen, a terminal multiplexer similar to tmux, and allows you to run both `bsb` and `ava` watchers in a single terminal. Try installing `screen` and run `npm run watch:screen`.

**Comment by dorafmon** [comment]
Cool, I use tmux myself but I never tried screen.

**Comment by glennsl** [comment]
I couldn't figure out how to get tmux to do what I wanted, but with screen it was easy, just needed the `.screenrc`

**Pull request merged**